src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/AsyncAnalyzerUtilities.cs

```diff
@@ -0,0 +1,116 @@
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Azure.ClientSdk.Analyzers 
+{
+    internal sealed class AsyncAnalyzerUtilities 
+    {
+        private readonly Compilation _compilation;
+
+        public INamedTypeSymbol TaskTypeSymbol { get; }
+        public INamedTypeSymbol TaskOfTTypeSymbol { get; }
+        public INamedTypeSymbol ValueTaskTypeSymbol { get; }
+        public INamedTypeSymbol ValueTaskOfTTypeSymbol { get; }
+        public INamedTypeSymbol AsyncDisposableSymbol { get; }
+        public INamedTypeSymbol AsyncEnumerableOfTTypeSymbol { get; }
+        public INamedTypeSymbol TaskAsyncEnumerableExtensionsSymbol { get; }
+
+        public AsyncAnalyzerUtilities(Compilation compilation) 
+        {
+            _compilation = compilation;
+            TaskTypeSymbol = compilation.GetTypeByMetadataName(typeof(Task).FullName);
+            TaskOfTTypeSymbol = compilation.GetTypeByMetadataName(typeof(Task<>).FullName);
+            ValueTaskTypeSymbol = compilation.GetTypeByMetadataName(typeof(ValueTask).FullName);
+            ValueTaskOfTTypeSymbol = compilation.GetTypeByMetadataName(typeof(ValueTask<>).FullName);
+            AsyncDisposableSymbol = compilation.GetTypeByMetadataName(typeof(IAsyncDisposable).FullName);
+            AsyncEnumerableOfTTypeSymbol = compilation.GetTypeByMetadataName(typeof(IAsyncEnumerable<>).FullName);
+            TaskAsyncEnumerableExtensionsSymbol = compilation.GetTypeByMetadataName(typeof(TaskAsyncEnumerableExtensions).FullName);
+        }
+
+        public bool IsConfigureAwait(SyntaxNode node) =>
+            node is InvocationExpressionSyntax invocation &&
+            invocation.Expression is MemberAccessExpressionSyntax memberAccess &&
+            memberAccess.Name.Identifier.Text == nameof(Task.ConfigureAwait);
+
+        public bool IsConfigureAwait(IMethodSymbol method)
+        {
+            if (method.Name == nameof(Task.ConfigureAwait) && method.Parameters.Length == 1 && IsTaskType(method.ReceiverType))
+            {
+                return true;
+            }
+
+            if (method.Name == nameof(TaskAsyncEnumerableExtensions.ConfigureAwait) && method.Parameters.Length == 2 && Equals(method.ReceiverType, TaskAsyncEnumerableExtensionsSymbol))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        public bool IsAsyncDisposableType(ITypeSymbol type)
+        {
+            if (type == null)
+            {
+                return false;
+            }
+
+            if (Equals(AsyncDisposableSymbol, type))
+            {
+                return true;
+            }
+
+            return type.AllInterfaces.Any(candidate => Equals(AsyncDisposableSymbol, candidate));
+        }
+
+        public bool IsAsyncEnumerableType(ITypeSymbol type)
+        {
+            if (type == null)
+            {
+                return false;
+            }
+
+            if (Equals(AsyncEnumerableOfTTypeSymbol, type.OriginalDefinition))
```

Comment:
Why do we use `.OriginalDefinition` here but not above?

Reply:
`AsyncEnumerable<T>` is a generic, and `type` is concrete. `AsyncDisposable` is not generic.

Reply:
Found a way to fix it by using `ConstructedFrom`.